File: backend/models/contact_message_model.go
package models

import (
	"fmt"
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// ContactMessageStatus represents the status of a contact message
type ContactMessageStatus string

const (
	ContactStatusNew      ContactMessageStatus = "new"      // Not yet reviewed
	ContactStatusRead     ContactMessageStatus = "read"     // Admin has viewed it
	ContactStatusReplied  ContactMessageStatus = "replied"  // Response sent
	ContactStatusArchived ContactMessageStatus = "archived" // Archived by admin
)

// ContactMessageCategory represents the category of a contact message
type ContactMessageCategory string

const (
	ContactCategoryGeneral  ContactMessageCategory = "general"
	ContactCategorySupport  ContactMessageCategory = "support"
	ContactCategorySales    ContactMessageCategory = "sales"
	ContactCategoryFeedback ContactMessageCategory = "feedback"
	ContactCategoryOther    ContactMessageCategory = "other"
)

// ContactMessage represents a contact form submission from public users
type ContactMessage struct {
	ID primitive.ObjectID `bson:"_id,omitempty" json:"id"`

	// Sender information
	Name  string `bson:"name" json:"name" binding:"required,min=2,max=100"`
	Email string `bson:"email" json:"email" binding:"required,email,max=100"`
	Phone string `bson:"phone,omitempty" json:"phone,omitempty"`

	// Message content
	Subject  string                 `bson:"subject" json:"subject" binding:"required,min=5,max=200"`
	Message  string                 `bson:"message" json:"message" binding:"required,min=10,max=5000"`
	Category ContactMessageCategory `bson:"category" json:"category"`

	// Status and tracking
	Status ContactMessageStatus `bson:"status" json:"status"`
	IsRead bool                 `bson:"isRead" json:"isRead"`

	// Admin response
	ReplyMessage string              `bson:"replyMessage,omitempty" json:"replyMessage,omitempty"`
	RepliedBy    *primitive.ObjectID `bson:"repliedBy,omitempty" json:"repliedBy,omitempty"`
	RepliedAt    *time.Time          `bson:"repliedAt,omitempty" json:"repliedAt,omitempty"`

	// Admin notes (internal)
	AdminNotes string              `bson:"adminNotes,omitempty" json:"adminNotes,omitempty"`
	AssignedTo *primitive.ObjectID `bson:"assignedTo,omitempty" json:"assignedTo,omitempty"`

	// Security
	IPAddress string `bson:"ipAddress,omitempty" json:"ipAddress,omitempty"`
	UserAgent string `bson:"userAgent,omitempty" json:"userAgent,omitempty"`

	// Timestamps
	CreatedAt time.Time  `bson:"createdAt" json:"createdAt"`
	UpdatedAt time.Time  `bson:"updatedAt" json:"updatedAt"`
	ReadAt    *time.Time `bson:"readAt,omitempty" json:"readAt,omitempty"`
}

// ContactMessageStats represents statistics for contact messages
type ContactMessageStats struct {
	TotalCount   int64 `json:"totalCount"`
	UnreadCount  int64 `json:"unreadCount"`
	NewCount     int64 `json:"newCount"`
	RepliedCount int64 `json:"repliedCount"`
}

// Validate validates contact message fields
func (cm *ContactMessage) Validate() error {
	if cm.Name == "" {
		return fmt.Errorf("name is required")
	}
	if len(cm.Name) < 2 || len(cm.Name) > 100 {
		return fmt.Errorf("name must be between 2 and 100 characters")
	}
	if cm.Email == "" {
		return fmt.Errorf("email is required")
	}
	if cm.Subject == "" {
		return fmt.Errorf("subject is required")
	}
	if len(cm.Subject) < 5 || len(cm.Subject) > 200 {
		return fmt.Errorf("subject must be between 5 and 200 characters")
	}
	if cm.Message == "" {
		return fmt.Errorf("message is required")
	}
	if len(cm.Message) < 10 || len(cm.Message) > 5000 {
		return fmt.Errorf("message must be between 10 and 5000 characters")
	}
	return nil
}

// SetDefaults sets default values for a new contact message
func (cm *ContactMessage) SetDefaults() {
	if cm.ID.IsZero() {
		cm.ID = primitive.NewObjectID()
	}
	now := time.Now()
	cm.CreatedAt = now
	cm.UpdatedAt = now
	cm.Status = ContactStatusNew
	cm.IsRead = false

	// Default category if not set
	if cm.Category == "" {
		cm.Category = ContactCategoryGeneral
	}
}

// MarkAsRead marks the message as read
func (cm *ContactMessage) MarkAsRead() {
	now := time.Now()
	cm.IsRead = true
	cm.ReadAt = &now
	cm.UpdatedAt = now
	if cm.Status == ContactStatusNew {
		cm.Status = ContactStatusRead
	}
}

// SetReply sets the admin reply for the message
func (cm *ContactMessage) SetReply(adminID primitive.ObjectID, replyMessage string) {
	now := time.Now()
	cm.ReplyMessage = replyMessage
	cm.RepliedBy = &adminID
	cm.RepliedAt = &now
	cm.Status = ContactStatusReplied
	cm.UpdatedAt = now
}

// ValidCategories returns the list of valid contact message categories
func ValidCategories() []string {
	return []string{
		string(ContactCategoryGeneral),
		string(ContactCategorySupport),
		string(ContactCategorySales),
		string(ContactCategoryFeedback),
		string(ContactCategoryOther),
	}
}

// ValidStatuses returns the list of valid contact message statuses
func ValidStatuses() []string {
	return []string{
		string(ContactStatusNew),
		string(ContactStatusRead),
		string(ContactStatusReplied),
		string(ContactStatusArchived),
	}
}

// IsValidCategory checks if a category is valid
func IsValidCategory(category string) bool {
	for _, c := range ValidCategories() {
		if c == category {
			return true
		}
	}
	return false
}

// IsValidStatus checks if a status is valid
func IsValidStatus(status string) bool {
	for _, s := range ValidStatuses() {
		if s == status {
			return true
		}
	}
	return false
}

File: street_core/lib/data/models/contact_message_model.dart
/// Model for contact messages sent through the contact form
class ContactMessage {

  ContactMessage({
    this.id,
    this.name,
    this.email,
    this.phone,
    this.subject,
    this.message,
    this.category,
    this.status,
    this.isReadFlag = false,
    this.createdAt,
    this.updatedAt,
    this.readAt,
    this.repliedAt,
    this.repliedBy,
    this.replyMessage,
    this.adminNotes,
    this.assignedTo,
  });

  factory ContactMessage.fromJson(Map<String, dynamic> json) {
    return ContactMessage(
      id: json['id'] ?? json['_id'],
      name: json['name'],
      email: json['email'],
      phone: json['phone'],
      subject: json['subject'],
      message: json['message'],
      category: json['category'] ?? 'general',
      status: json['status'] ?? 'new',
      isReadFlag: json['isRead'] ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'])
          : null,
      readAt:
          json['readAt'] != null ? DateTime.tryParse(json['readAt']) : null,
      repliedAt: json['repliedAt'] != null
          ? DateTime.tryParse(json['repliedAt'])
          : null,
      repliedBy: json['repliedBy'],
      replyMessage: json['replyMessage'],
      adminNotes: json['adminNotes'],
      assignedTo: json['assignedTo'],
    );
  }
  final String? id;
  final String? name;
  final String? email;
  final String? phone;
  final String? subject;
  final String? message;
  final String? category; // 'general', 'support', 'sales', 'feedback', 'other'
  final String? status; // 'new', 'read', 'replied', 'archived'
  final bool isReadFlag;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final DateTime? readAt;
  final DateTime? repliedAt;
  final String? repliedBy;
  final String? replyMessage;
  final String? adminNotes;
  final String? assignedTo;

  Map<String, dynamic> toJson() => {
        if (id != null) 'id': id,
        if (name != null) 'name': name,
        if (email != null) 'email': email,
        if (phone != null) 'phone': phone,
        if (subject != null) 'subject': subject,
        if (message != null) 'message': message,
        if (category != null) 'category': category,
        if (status != null) 'status': status,
        'isRead': isReadFlag,
        if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
        if (updatedAt != null) 'updatedAt': updatedAt!.toIso8601String(),
        if (readAt != null) 'readAt': readAt!.toIso8601String(),
        if (repliedAt != null) 'repliedAt': repliedAt!.toIso8601String(),
        if (repliedBy != null) 'repliedBy': repliedBy,
        if (replyMessage != null) 'replyMessage': replyMessage,
        if (adminNotes != null) 'adminNotes': adminNotes,
        if (assignedTo != null) 'assignedTo': assignedTo,
      };

  /// For form submission (only send required fields)
  Map<String, dynamic> toSubmitJson() => {
        'name': name,
        'email': email,
        if (phone != null && phone!.isNotEmpty) 'phone': phone,
        'subject': subject,
        'message': message,
        if (category != null) 'category': category,
      };

  bool get isNew => status == 'new';
  bool get isRead =>
      isReadFlag ||
      status == 'read' ||
      status == 'replied' ||
      status == 'archived';
  bool get isReplied => status == 'replied';
  bool get isArchived => status == 'archived';

  ContactMessage copyWith({
    String? id,
    String? name,
    String? email,
    String? phone,
    String? subject,
    String? message,
    String? category,
    String? status,
    bool? isReadFlag,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? readAt,
    DateTime? repliedAt,
    String? repliedBy,
    String? replyMessage,
    String? adminNotes,
    String? assignedTo,
  }) {
    return ContactMessage(
      id: id ?? this.id,
      name: name ?? this.name,
      email: email ?? this.email,
      phone: phone ?? this.phone,
      subject: subject ?? this.subject,
      message: message ?? this.message,
      category: category ?? this.category,
      status: status ?? this.status,
      isReadFlag: isReadFlag ?? this.isReadFlag,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      readAt: readAt ?? this.readAt,
      repliedAt: repliedAt ?? this.repliedAt,
      repliedBy: repliedBy ?? this.repliedBy,
      replyMessage: replyMessage ?? this.replyMessage,
      adminNotes: adminNotes ?? this.adminNotes,
      assignedTo: assignedTo ?? this.assignedTo,
    );
  }
}

/// Stats model for contact messages
class ContactMessageStats {

  ContactMessageStats({
    this.totalCount = 0,
    this.unreadCount = 0,
    this.newCount = 0,
    this.repliedCount = 0,
  });

  factory ContactMessageStats.fromJson(Map<String, dynamic> json) {
    return ContactMessageStats(
      totalCount: json['totalCount'] ?? 0,
      unreadCount: json['unreadCount'] ?? 0,
      newCount: json['newCount'] ?? 0,
      repliedCount: json['repliedCount'] ?? 0,
    );
  }
  final int totalCount;
  final int unreadCount;
  final int newCount;
  final int repliedCount;
}
